realm/realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -897,24 +897,139 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeNotEqual__J_3JLja
     TableQuery_StringPredicate(env, nativeQueryPtr, columnIndexes, value, caseSensitive, StringNotEqual);
 }
 
-JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeBeginsWith(
+JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeBeginsWith__J_3JLjava_lang_String_2Z(
     JNIEnv* env, jobject, jlong nativeQueryPtr, jlongArray columnIndexes, jstring value, jboolean caseSensitive)
 {
     TableQuery_StringPredicate(env, nativeQueryPtr, columnIndexes, value, caseSensitive, StringBeginsWith);
 }
 
-JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEndsWith(
+JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEndsWith__J_3JLjava_lang_String_2Z(
     JNIEnv* env, jobject, jlong nativeQueryPtr, jlongArray columnIndexes, jstring value, jboolean caseSensitive)
 {
     TableQuery_StringPredicate(env, nativeQueryPtr, columnIndexes, value, caseSensitive, StringEndsWith);
 }
 
-JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeContains(
+JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeContains__J_3JLjava_lang_String_2Z(
     JNIEnv* env, jobject, jlong nativeQueryPtr, jlongArray columnIndexes, jstring value, jboolean caseSensitive)
 {
     TableQuery_StringPredicate(env, nativeQueryPtr, columnIndexes, value, caseSensitive, StringContains);
 }
 
+// Binary
+
+enum BinaryPredicate {
+    BinaryEqual,
+    BinaryNotEqual,
+    BinaryContains,
+    BinaryBeginsWith,
+    BinaryEndsWith
+};
+
+static void TableQuery_BinaryPredicate(JNIEnv *env, jlong nativeQueryPtr, jlongArray columnIndices, jbyteArray value, BinaryPredicate predicate) {
+    JniLongArray arr(env, columnIndices);
+    jsize arr_len = arr.len();
+    try {
+        BinaryData value2;
+        if (value == NULL) {
+            if (!TBL_AND_COL_NULLABLE(env, getTableByArray(nativeQueryPtr, arr).get(), arr[arr_len-1])) {
+                return;
+            }
+            value2 = BinaryData();
+        }
+        else {
+            jbyte* bytePtr = env->GetByteArrayElements(value, NULL);
+            if (!bytePtr) {
+                ThrowException(env, IllegalArgument, "binaryPredicate");
+                return;
+            }
+            size_t dataLen = S(env->GetArrayLength(value));
+            value2 = BinaryData(reinterpret_cast<char*>(bytePtr), dataLen);
+            env->ReleaseByteArrayElements(value, bytePtr, 0);
```

Comment:
In `tbl_nativeDoByteArray`, `bytePtr` is not used after invoking `ReleaseByteArrayElements`.
`(pTable->*doBinary)( S(columnIndex), S(rowIndex), realm::BinaryData(reinterpret_cast<char*>(bytePtr), dataLen));` copies bytes to its own array and does not access the pointer from `BinaryData` after returning the function (the code is in the core  and not yet public though).

The official document of `ReleaseBooleanArrayElements` clearly says that this function frees `bytePtr` if the `mode` is `0` or `JNI_ABORT`.
http://docs.oracle.com/javase/1.5.0/docs/guide/jni/spec/functions.html#wp1544

I think `ReleaseByteArrayElements` must be at the end of this function.


And I confirmed that `Q(nativeQueryPtr)->equal(S(arr[0]), value2);` and `Q(nativeQueryPtr)->not_equal(S(arr[0]), value2);` in `TableQuery_BinaryPredicate` copy byte array in `value2`, 
but I'm not sure about `table_ref->column<Binary>(size_t(arr[arr_len-1])) == value2` and `table_ref->column<Binary>(size_t(arr[arr_len-1])) != value2`. What do you think @kneth ?